### PR TITLE
[6.0] [Mangler] Mangle function type for invalid enum element

### PR DESCRIPTION
**Pull request by hamishknight**
*6.0 cherry-pick of #74288*

- Explanation: Fixes an assertion failure for an invalid enum case when generating a USR
- Scope: Affects USR generation for invalid enum cases
- Issue: rdar://129065620
- Risk: Low
- Testing: Added tests to test suite
- Reviewer: Alex Hoppen

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Review by slavapestov — APPROVED** [review]

File: lib/AST/ASTMangler.cpp [hunk 1]
```diff
@@ -3838,7 +3838,7 @@ CanType ASTMangler::getDeclTypeForMangling(
 
   auto &C = decl->getASTContext();
   if (decl->isInvalid()) {
-    if (isa<AbstractFunctionDecl>(decl)) {
+    if (isa<AbstractFunctionDecl>(decl) || isa<EnumElementDecl>(decl)) {
```
Comment (slavapestov) [inline lib/AST/ASTMangler.cpp:3841]: Maybe subscripts as well?

File: lib/AST/ASTMangler.cpp [hunk 2]
```diff
@@ -3838,7 +3838,7 @@ CanType ASTMangler::getDeclTypeForMangling(
 
   auto &C = decl->getASTContext();
   if (decl->isInvalid()) {
```
Comment (slavapestov) [inline lib/AST/ASTMangler.cpp:3840]: Move this up here: `Type ty = decl->getInterfaceType()->getReferenceStorageReferent();` and replace the `isInvalid()` check with `if (ty->hasError())`

**Review by slavapestov — APPROVED** [review]
Two minor nitpicks

**Comment by slavapestov** [comment]
Sorry, I meant to post the second review on the `main` PR. 6.0 PR doesn't need any changes!

File: lib/AST/ASTMangler.cpp (hunk 1 above)
Comment (hamishknight) [reply lib/AST/ASTMangler.cpp:3841]: Looks like the demangler doesn't currently require a function type for a subscript, but yeah let's also handle that, good catch!

**Review by bnbarham — APPROVED** [review]

**Comment by hamishknight** [comment]
@swift-ci please test

**Pull request merged**